axelor-contract/src/main/java/com/axelor/apps/contract/service/ContractSalePurchaseOrderGenerationImpl.java

```diff
@@ -0,0 +1,195 @@
+  protected AnalyticMoveLineRepository analyticMoveLineRepo;
+  protected SaleOrderCreateService saleOrderCreateService;
+  protected SaleOrderComputeService saleOrderComputeService;
+  protected PurchaseOrderService purchaseOrderService;
+
+  @Inject
+  public ContractSalePurchaseOrderGenerationImpl(
+      AppBaseService appBaseService,
+      SaleOrderRepository saleOrderRepository,
+      PurchaseOrderRepository purchaseOrderRepository,
+      AnalyticMoveLineRepository analyticMoveLineRepo,
+      SaleOrderCreateService saleOrderCreateService,
+      SaleOrderComputeService saleOrderComputeService,
+      PurchaseOrderService purchaseOrderService) {
+    this.appBaseService = appBaseService;
+    this.saleOrderRepository = saleOrderRepository;
+    this.purchaseOrderRepository = purchaseOrderRepository;
+    this.analyticMoveLineRepo = analyticMoveLineRepo;
+    this.saleOrderCreateService = saleOrderCreateService;
+    this.saleOrderComputeService = saleOrderComputeService;
+    this.purchaseOrderService = purchaseOrderService;
+  }
+
+  @Override
+  @Transactional(rollbackOn = {Exception.class})
+  public SaleOrder generateSaleOrder(Contract contract) throws AxelorException {
+    Partner clientPartner = contract.getPartner();
+
+    SaleOrder saleOrder = saleOrderCreateService.createSaleOrder(contract.getCompany());
+    saleOrder.setCurrency(contract.getCurrency());
+    saleOrder.setClientPartner(clientPartner);
+    saleOrder.setInvoicedPartner(contract.getInvoicedPartner());
+    saleOrder.setPaymentMode(contract.getCurrentContractVersion().getPaymentMode());
+    saleOrder.setPaymentCondition(contract.getInvoicedPartner().getPaymentCondition());
+    saleOrder.setContract(contract);
+
+    if (appBaseService.getAppBase().getActivatePartnerRelations()) {
+      saleOrder.setDeliveredPartner(clientPartner);
+    }
+
+    for (ContractLine contractLine : contract.getCurrentContractVersion().getContractLineList()) {
+      createSaleOrderLineFromContractLine(contractLine, saleOrder);
+    }
+
+    for (ContractLine contractLine : contract.getAdditionalBenefitContractLineList()) {
+      createSaleOrderLineFromContractLine(contractLine, saleOrder);
+    }
+    saleOrderComputeService.computeSaleOrder(saleOrder);
+    return saleOrderRepository.save(saleOrder);
+  }
+
+  @Override
+  @Transactional(rollbackOn = {Exception.class})
+  public PurchaseOrder generatePurchaseOrder(Contract contract) throws AxelorException {
+
+    PurchaseOrder purchaseOrder = new PurchaseOrder();
+    purchaseOrder.setStatusSelect(PurchaseOrderRepository.STATUS_DRAFT);
+    purchaseOrder.setCompany(contract.getCompany());
+    purchaseOrder.setSupplierPartner(contract.getPartner());
+    purchaseOrder.setCurrency(contract.getCurrency());
+    purchaseOrder.setPaymentMode(contract.getCurrentContractVersion().getPaymentMode());
+    purchaseOrder.setPaymentCondition(contract.getPartner().getPaymentCondition());
+    purchaseOrder.setContract(contract);
```

Comment:
please call PurchaseOrderService#createPurchaseOrder instead of calling directly the constructor